src/radical/pilot/agent/agent_0.py

```diff
@@ -340,76 +349,53 @@ def _start_services(self):
-                'executable'    : '/bin/sh',
-                'arguments'     : [bs_name, 'services']
-            }).as_dict(),
-            'slots': {'ranks'   : [{'node_name': nodes[0]['node_name'],
-                                    'node_id'  : nodes[0]['node_id'],
-                                    'core_map' : [[0]],
-                                    'gpu_map'  : [],
-                                    'lfs'      : 0,
-                                    'mem'      : 0}]}
-        }
-
-        launcher = self._rm.find_launcher(service_task)
-        if not launcher:
-            raise RuntimeError('no launch method found for sub agent')
-
-        # FIXME: set RP environment (as in Popen Executor)
-
-        tmp  = '#!/bin/sh\n\n'
-        tmp += 'export RP_PILOT_SANDBOX="%s"\n\n' % self._pwd
-        cmds = launcher.get_launcher_env()
-        for cmd in cmds:
-            tmp += '%s || exit 1\n' % cmd
-        cmds = launcher.get_launch_cmd(service_task, ex_name)
-        tmp += '%s\nexit $?\n\n' % '\n'.join(cmds)
-
-        with ru.ru_open(ls_name, 'w') as fout:
-            fout.write(tmp)
-
-        tmp  = '#!/bin/sh\n\n'
-        tmp += '. ./env/service.env\n'
-        tmp += '/bin/sh -l ./services\n\n'
-
-        with ru.ru_open(ex_name, 'w') as fout:
-            fout.write(tmp)
-
-
-        # make sure scripts are executable
-        st_l = os.stat(ls_name)
-        st_e = os.stat(ex_name)
-        os.chmod(ls_name, st_l.st_mode | stat.S_IEXEC)
-        os.chmod(ex_name, st_e.st_mode | stat.S_IEXEC)
-
-        # spawn the sub-agent
-        cmdline = './%s' % ls_name
-
-        self._log.info('create services: %s', cmdline)
-        ru.sh_callout_bg(cmdline, stdout='services.out', stderr='services.err')
-
-        self._log.debug('services started done')
-
+        self._log.info('Starting the agent services')
+        service_descriptions = self._cfg.services
+        services = list()
+
+        for service_desc in service_descriptions:
+            task = dict()
+            task['origin']            = 'agent'
+            task['description']       = TaskDescription(service_desc).as_dict()
+            task['state']             = rps.AGENT_STAGING_INPUT_PENDING
+            task['status']            = 'NEW'
+            task['type']              = 'service_task'
+            task['uid']               = ru.generate_id('service.%(item_counter)04d',
+                                             ru.ID_CUSTOM, ns=self._cfg.sid)
+            task['pilot_sandbox']     = self._cfg.pilot_sandbox
+            task['task_sandbox']      = self._cfg.pilot_sandbox + task['uid'] + '/'
+            task['task_sandbox_path'] = self._cfg.pilot_sandbox + task['uid'] + '/'
+            task['session_sandbox']   = self._cfg.session_sandbox
+            task['resource_sandbox']  = self._cfg.resource_sandbox
+            task['pilot']             = self._cfg.pid
+            self._service_task_ids.append(task['uid'])
```

Comment:
A recent commit removed the (commented) `task['resources']` entry.  We will need that for post mortem run analysis - please re-add and uncomment.

Reply:
Yes, I did remove it. Didn't realize it.

Reply:
I am taking `service_desc['ranks']` without default values. Is it assured that this key is always present ?

Reply:
Yes, that is ensured by the `TaskDescription` class.

Reply:
I just realized that you pull `ranks` from the `service_description` dict - sorry for missing your point.  You probably want to pull it from the `TakDescription` instance you create a couple of lines above.  The whole purpose of that step is to fill defaults and do sanity checks (otherwise we could just pass the original dict).  So once that description is created, please use it for all look-ups, not the original `service_desc` dict.

Reply:
Yes now i get. From the start I had this confusion within me. Thanks for clarifying.